File: src/game/activity/transition.ts
import type { GameDatabase } from '../data/types'
import type { ActivityTransition, PlayerSave } from '../save/types'
import { isDeathPaused } from '../combat/engine'
import type { RandomFn } from './pools'
import { cancelProductionActivity, beginProductionQueue } from '../production/engine'
import { isStandardProductionActivity } from '../production/recipes'
import { unequipSlot } from '../equipment/loadout'
import { withRecalculatedVitals } from '../equipment/vitals'
import {
  beginActivitySave,
  clearActivitySave,
  generateNextAction,
  getActivity,
  isBlessingActivity,
  validateActivityStart,
} from './engine'
import { requirementsForEntity } from './requirements'
import {
  forcedHostileActivity,
  HOSTILE_ACTIVITY_LOCK_REASON,
  HOSTILE_ACTIVITY_START_REASON,
  locationIsHostileFor,
} from '../world/hostility'

export function clearActivityTransition(save: PlayerSave): PlayerSave {
  if (!save.activityTransition) return save
  return { ...save, activityTransition: null }
}

export function hasRunningPrimaryActivity(save: PlayerSave): boolean {
  return Boolean(save.currentActivityId || save.productionRecipeId)
}

/** Immediately stop the current Primary Activity (refunds remaining production materials). */
export function stopPrimaryActivityNow(
  db: GameDatabase,
  save: PlayerSave,
  nowMs: number = Date.now(),
): PlayerSave {
  // Death pause keeps the Primary Activity until recovery finishes.
  if (isDeathPaused(save, nowMs)) return save

  let next = clearActivityTransition(save)
  if (next.productionRecipeId) {
    next = cancelProductionActivity(db, next)
  } else if (next.currentActivityId) {
    next = clearActivitySave(next, nowMs)
  }
  return next
}

/**
 * Travel interrupt: hard-stop the current Primary Activity immediately.
 * No activity-change cooldown — death pause still blocks travel in the UI/hostility layer.
 */
export function beginTravelActivityChange(
  db: GameDatabase,
  save: PlayerSave,
  nowMs: number = Date.now(),
): PlayerSave {
  if (isDeathPaused(save, nowMs)) return save
  if (!hasRunningPrimaryActivity(save) && !save.activityTransition) return save
  return stopPrimaryActivityNow(db, save, nowMs)
}

function startPoolActivityNow(
  db: GameDatabase,
  save: PlayerSave,
  activityId: string,
  nowMs: number,
  random: RandomFn,
): PlayerSave {
  const nowIso = new Date(nowMs).toISOString()
  const cleared = clearActivityTransition(save)
  const started = beginActivitySave(cleared, activityId, nowIso)
  const generated = generateNextAction(db, started, activityId, random, nowMs)
  return generated ? generated.save : started
}

function startProductionNow(
  db: GameDatabase,
  save: PlayerSave,
  activityId: string,
  recipeId: string,
  quantity: number,
  nowMs: number,
): PlayerSave | { ok: false; reason: string } {
  const nowIso = new Date(nowMs).toISOString()
  const cleared = clearActivityTransition(save)
  const started = beginActivitySave(cleared, activityId, nowIso)
  const queued = beginProductionQueue(db, started, activityId, recipeId, quantity, nowMs)
  if (!queued.ok) return queued
  return queued.save
}

function hostileStartBlocked(
  db: GameDatabase,
  save: PlayerSave,
  activityId: string,
): string | null {
  if (!locationIsHostileFor(db, save)) return null
  const threatened = forcedHostileActivity(db, save, save.currentLocationId)
  if (threatened && threatened['Activity ID'] === activityId) return null
  return HOSTILE_ACTIVITY_START_REASON
}

/** Unequip slots an activity requires empty before start validation. */
export function unequipEmptySlotRequirements(
  db: GameDatabase,
  save: PlayerSave,
  activityId: string,
): { ok: true; save: PlayerSave } | { ok: false; reason: string } {
  let next = save
  for (const requirement of requirementsForEntity(db, 'Activity', activityId)) {
    if (requirement['Requirement Type'] !== 'Empty Slot') continue
    const slotId = String(requirement['Reference ID / Value'] ?? '')
    if (!slotId) continue
    const equipped = next.equipment.slots[slotId]
    if (!equipped?.itemId || equipped.quantity <= 0) continue
    const result = unequipSlot(next, slotId)
    if (!result.ok) return result
    next = result.save
  }
  return { ok: true, save: next }
}

function restoreFullHealth(db: GameDatabase, save: PlayerSave): PlayerSave {
  const next = withRecalculatedVitals(db, save)
  return { ...next, currentHp: next.maxHp }
}

/** Start or replace a pool activity immediately. Death pause still blocks. */
export function requestActivityStart(
  db: GameDatabase,
  save: PlayerSave,
  activityId: string,
  nowMs: number = Date.now(),
  random: RandomFn = Math.random,
): { ok: true; save: PlayerSave } | { ok: false; reason: string } {
  if (isDeathPaused(save, nowMs)) {
    return { ok: false, reason: 'Cannot change activities while recovering from defeat.' }
  }
  const hostile = hostileStartBlocked(db, save, activityId)
  if (hostile) return { ok: false, reason: hostile }

  const unequipped = unequipEmptySlotRequirements(db, save, activityId)
  if (!unequipped.ok) return unequipped
  let next = unequipped.save

  const validation = validateActivityStart(db, next, activityId)
  if (!validation.ok) return validation

  if (
    next.currentActivityId === activityId &&
    !next.productionRecipeId &&
    !next.activityTransition
  ) {
    return { ok: true, save: clearActivityTransition(next) }
  }

  if (hasRunningPrimaryActivity(next) || next.activityTransition) {
    next = stopPrimaryActivityNow(db, next, nowMs)
  }

  if (isBlessingActivity(getActivity(db, activityId))) {
    return { ok: true, save: restoreFullHealth(db, next) }
  }

  return { ok: true, save: startPoolActivityNow(db, next, activityId, nowMs, random) }
}

/** Start Standard Production immediately, replacing any running Primary Activity. */
export function requestProductionStart(
  db: GameDatabase,
  save: PlayerSave,
  activityId: string,
  recipeId: string,
  quantity: number,
  nowMs: number = Date.now(),
): { ok: true; save: PlayerSave } | { ok: false; reason: string } {
  if (isDeathPaused(save, nowMs)) {
    return { ok: false, reason: 'Cannot change activities while recovering from defeat.' }
  }
  if (locationIsHostileFor(db, save)) {
    return { ok: false, reason: HOSTILE_ACTIVITY_START_REASON }
  }

  const validation = validateActivityStart(db, save, activityId)
  if (!validation.ok) return validation

  let next = save
  if (hasRunningPrimaryActivity(save) || save.activityTransition) {
    next = stopPrimaryActivityNow(db, save, nowMs)
  }

  const started = startProductionNow(db, next, activityId, recipeId, quantity, nowMs)
  if ('ok' in started && started.ok === false) return started
  return { ok: true, save: started as PlayerSave }
}

/** Stop the current Primary Activity immediately. */
export function requestActivityStop(
  db: GameDatabase,
  save: PlayerSave,
  nowMs: number = Date.now(),
): { ok: true; save: PlayerSave } | { ok: false; reason: string } {
  if (isDeathPaused(save, nowMs)) {
    return { ok: false, reason: 'Cannot change activities while recovering from defeat.' }
  }
  if (locationIsHostileFor(db, save)) {
    return { ok: false, reason: HOSTILE_ACTIVITY_LOCK_REASON }
  }
  if (!hasRunningPrimaryActivity(save) && !save.activityTransition) {
    return { ok: false, reason: 'No activity is running.' }
  }

  return { ok: true, save: stopPrimaryActivityNow(db, save, nowMs) }
}

function applyFollowUpStart(
  db: GameDatabase,
  save: PlayerSave,
  transition: ActivityTransition,
  nowMs: number,
  random: RandomFn,
): PlayerSave {
  const followUp = transition.followUpActivityId
  if (!followUp) return save

  if (transition.productionRecipeId && transition.productionQuantity) {
    const started = startProductionNow(
      db,
      save,
      followUp,
      transition.productionRecipeId,
      transition.productionQuantity,
      nowMs,
    )
    if ('ok' in started && started.ok === false) return save
    return started as PlayerSave
  }

  const activity = getActivity(db, followUp)
  if (activity && isStandardProductionActivity(db, activity)) {
    return save
  }

  return startPoolActivityNow(db, save, followUp, nowMs, random)
}

/**
 * Clear legacy activity-change delays from older saves (apply any follow-up immediately).
 * New gameplay no longer queues these transitions.
 */
export function resolveActivityTransitions(
  db: GameDatabase,
  save: PlayerSave,
  nowMs: number = Date.now(),
  random: RandomFn = Math.random,
): PlayerSave {
  const transition = save.activityTransition
  if (!transition) return save

  const next = stopPrimaryActivityNow(db, save, nowMs)
  if (transition.kind === 'starting') {
    return applyFollowUpStart(
      db,
      next,
      { ...transition, followUpActivityId: transition.activityId },
      nowMs,
      random,
    )
  }
  return applyFollowUpStart(db, next, transition, nowMs, random)
}

File: packages/ik_rules/lib/src/activity/transition.dart
import 'package:ik_content/ik_content.dart';

import '../combat/engine.dart';
import '../js_compat.dart';
import '../production/engine.dart';
import '../production/recipes.dart';
import '../rng/mulberry32.dart';
import '../save/generated/save_models.dart';
import '../time.dart';
import 'engine.dart';
import '../world/hostility.dart';

PlayerSave clearActivityTransition(PlayerSave save) {
  if (save.activityTransition == null) return save;
  return save.copyWith(activityTransition: null);
}

bool hasRunningPrimaryActivity(PlayerSave save) {
  return isNotBlank(save.currentActivityId) || isNotBlank(save.productionRecipeId);
}

/// Immediately stops the current Primary Activity (refunds remaining production materials).
PlayerSave stopPrimaryActivityNow(GameDatabase db, PlayerSave save, num nowMs) {
  // Death pause keeps the Primary Activity until recovery finishes.
  if (isDeathPaused(save, nowMs)) return save;

  var next = clearActivityTransition(save);
  if (isNotBlank(next.productionRecipeId)) {
    next = cancelProductionActivity(db, next);
  } else if (isNotBlank(next.currentActivityId)) {
    next = clearActivitySave(next, nowMs);
  }
  return next;
}

/// Travel interrupt: hard-stops the current Primary Activity immediately.
///
/// No activity-change cooldown — death pause still blocks travel in the
/// UI/hostility layer.
PlayerSave beginTravelActivityChange(GameDatabase db, PlayerSave save, num nowMs) {
  if (isDeathPaused(save, nowMs)) return save;
  if (!hasRunningPrimaryActivity(save) && save.activityTransition == null) return save;
  return stopPrimaryActivityNow(db, save, nowMs);
}

PlayerSave _startPoolActivityNow(
  GameDatabase db,
  PlayerSave save,
  String activityId,
  num nowMs,
  RandomFn random,
) {
  final started = beginActivitySave(clearActivityTransition(save), activityId, isoFromMs(nowMs));
  return generateNextAction(db, started, activityId, random, nowMs)?.save ?? started;
}

/// Either the updated save or the reason the change was refused.
class ActivityChangeResult {
  const ActivityChangeResult.ok(this.save) : reason = null;

  const ActivityChangeResult.failed(this.reason) : save = null;

  bool get ok => reason == null;
  final PlayerSave? save;
  final String? reason;
}

ProductionQueueResult _startProductionNow(
  GameDatabase db,
  PlayerSave save,
  String activityId,
  String recipeId,
  num quantity,
  num nowMs,
) {
  final started = beginActivitySave(clearActivityTransition(save), activityId, isoFromMs(nowMs));
  return beginProductionQueue(db, started, activityId, recipeId, quantity, nowMs);
}

String? _hostileStartBlocked(GameDatabase db, PlayerSave save, String activityId) {
  if (!locationIsHostileFor(db, save)) return null;
  final threatened = forcedHostileActivity(db, save, save.currentLocationId);
  if (threatened != null && jsString(threatened.raw['Activity ID']) == activityId) {
    return null;
  }
  return hostileActivityStartReason;
}

/// Starts or replaces a pool activity immediately. Death pause still blocks.
ActivityChangeResult requestActivityStart(
  GameDatabase db,
  PlayerSave save,
  String activityId,
  num nowMs,
  RandomFn random,
) {
  if (isDeathPaused(save, nowMs)) {
    return const ActivityChangeResult.failed(
      'Cannot change activities while recovering from defeat.',
    );
  }
  final hostile = _hostileStartBlocked(db, save, activityId);
  if (hostile != null) return ActivityChangeResult.failed(hostile);

  final validation = validateActivityStart(db, save, activityId);
  if (!validation.ok) return ActivityChangeResult.failed(validation.reason);

  if (save.currentActivityId == activityId &&
      isBlank(save.productionRecipeId) &&
      save.activityTransition == null) {
    return ActivityChangeResult.ok(clearActivityTransition(save));
  }

  var next = save;
  if (hasRunningPrimaryActivity(save) || save.activityTransition != null) {
    next = stopPrimaryActivityNow(db, save, nowMs);
  }

  return ActivityChangeResult.ok(_startPoolActivityNow(db, next, activityId, nowMs, random));
}

/// Starts Standard Production immediately, replacing any running Primary Activity.
ActivityChangeResult requestProductionStart(
  GameDatabase db,
  PlayerSave save,
  String activityId,
  String recipeId,
  num quantity,
  num nowMs,
) {
  if (isDeathPaused(save, nowMs)) {
    return const ActivityChangeResult.failed(
      'Cannot change activities while recovering from defeat.',
    );
  }
  if (locationIsHostileFor(db, save)) {
    return const ActivityChangeResult.failed(hostileActivityStartReason);
  }

  final validation = validateActivityStart(db, save, activityId);
  if (!validation.ok) return ActivityChangeResult.failed(validation.reason);

  var next = save;
  if (hasRunningPrimaryActivity(save) || save.activityTransition != null) {
    next = stopPrimaryActivityNow(db, save, nowMs);
  }

  final started = _startProductionNow(db, next, activityId, recipeId, quantity, nowMs);
  if (!started.ok) return ActivityChangeResult.failed(started.reason);
  return ActivityChangeResult.ok(started.save);
}

/// Stops the current Primary Activity immediately.
ActivityChangeResult requestActivityStop(GameDatabase db, PlayerSave save, num nowMs) {
  if (isDeathPaused(save, nowMs)) {
    return const ActivityChangeResult.failed(
      'Cannot change activities while recovering from defeat.',
    );
  }
  if (locationIsHostileFor(db, save)) {
    return const ActivityChangeResult.failed(hostileActivityLockReason);
  }
  if (!hasRunningPrimaryActivity(save) && save.activityTransition == null) {
    return const ActivityChangeResult.failed('No activity is running.');
  }

  return ActivityChangeResult.ok(stopPrimaryActivityNow(db, save, nowMs));
}

PlayerSave _applyFollowUpStart(
  GameDatabase db,
  PlayerSave save,
  ActivityTransition transition,
  num nowMs,
  RandomFn random,
) {
  final followUp = transition.followUpActivityId;
  if (isBlank(followUp)) return save;

  final recipeId = transition.productionRecipeId;
  final quantity = transition.productionQuantity;
  if (isNotBlank(recipeId) && quantity != null && quantity != 0) {
    final started = _startProductionNow(db, save, followUp!, recipeId!, quantity, nowMs);
    return started.ok ? started.save! : save;
  }

  final activity = getActivity(db, followUp!);
  if (activity != null && isStandardProductionActivity(db, activity)) return save;

  return _startPoolActivityNow(db, save, followUp, nowMs, random);
}

/// Clears legacy activity-change delays from older saves, applying any follow-up
/// immediately. New gameplay no longer queues these transitions.
PlayerSave resolveActivityTransitions(
  GameDatabase db,
  PlayerSave save,
  num nowMs,
  RandomFn random,
) {
  final transition = save.activityTransition;
  if (transition == null) return save;

  final next = stopPrimaryActivityNow(db, save, nowMs);
  if (transition.kind == 'starting') {
    return _applyFollowUpStart(
      db,
      next,
      transition.copyWith(followUpActivityId: transition.activityId),
      nowMs,
      random,
    );
  }
  return _applyFollowUpStart(db, next, transition, nowMs, random);
}
